src/main/java/io/anserini/search/SearchShardedHnswDenseVectors.java

```diff
@@ -0,0 +1,237 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.anserini.search;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+import io.anserini.index.ShardInfo;
+import io.anserini.search.topicreader.TopicReader;
+import io.anserini.search.topicreader.Topics;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kohsuke.args4j.CmdLineException;
+import org.kohsuke.args4j.CmdLineParser;
+import org.kohsuke.args4j.Option;
+import org.kohsuke.args4j.ParserProperties;
+import org.kohsuke.args4j.spi.StringArrayOptionHandler;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.IntStream;
+
+/**
+ * Main entry point for sharded HNSW dense vector search.
+ */
+public final class SearchShardedHnswDenseVectors<K extends Comparable<K>> implements Runnable, Closeable {
+  private static final Logger LOG = LogManager.getLogger(SearchShardedHnswDenseVectors.class);
+
+  public static class Args extends HnswDenseSearcher.Args {
+    @Option(name = "-topics", metaVar = "[file]", handler = StringArrayOptionHandler.class, required = true, usage = "topics file")
+    public String[] topics;
+
+    @Option(name = "-output", metaVar = "[file]", required = true, usage = "output file")
+    public String output;
+
+    @Option(name = "-topicReader", usage = "TopicReader to use.")
+    public String topicReader = "JsonIntVector";
+
+    @Option(name = "-topicField", usage = "Topic field that should be used as the query.")
+    public String topicField = "vector";
+
+    @Option(name = "-hits", metaVar = "[number]", usage = "max number of hits to return")
+    public int hits = 1000;
+
+    @Option(name = "-runtag", metaVar = "[tag]", usage = "runtag")
+    public String runtag = "Anserini";
+
+    @Option(name = "-format", metaVar = "[output format]", usage = "Output format, default \"trec\", alternative \"msmarco\".")
+    public String format = "trec";
+
+    @Option(name = "-options", usage = "Print information about options.")
+    public Boolean options = false;
+  }
+
+  private final Args args;
+  private final List<HnswDenseSearcher<K>> searchers;
+  private final List<K> qids = new ArrayList<>();
+  private final List<String> queries = new ArrayList<>();
+  private final ShardInfo[] shards;
+  private final int threadsPerShard;
+
+  /*
+   * Constructor for sharded HNSW dense vector search.
+   * The caller must provide an identifier for the sharded index defined in ShardInfo.
+   * This identifier is used to retrieve the shards from ShardInfo.
+   */
+  public SearchShardedHnswDenseVectors(Args args) throws IOException {
```

Comment:
What if we just re-packaged the params and dispatched to `SearchHnswDenseVectors` in parallel? Wouldn't this reduce code duplication?